internal/generate/clusterserviceversion/clusterserviceversion_updaters.go

```diff
@@ -312,16 +332,81 @@ func mutatingToWebhookDescription(webhook admissionregv1.MutatingWebhook) operat
 		AdmissionReviewVersions: webhook.AdmissionReviewVersions,
 		ReinvocationPolicy:      webhook.ReinvocationPolicy,
 	}
+	if len(description.AdmissionReviewVersions) == 0 {
+		description.AdmissionReviewVersions = defaultAdmissionReviewVersions
+	}
+
 	if serviceRef := webhook.ClientConfig.Service; serviceRef != nil {
 		if serviceRef.Port != nil {
 			description.ContainerPort = *serviceRef.Port
 		}
-		description.DeploymentName = strings.TrimSuffix(serviceRef.Name, "-service")
+		description.DeploymentName = depName
+		if description.DeploymentName == "" {
+			description.DeploymentName = strings.TrimSuffix(serviceRef.Name, "-service")
+		}
 		description.WebhookPath = serviceRef.Path
 	}
 	return description
 }
 
+// findMatchingDeploymentAndServiceForWebhook matches a Service to a webhook's client config (if it uses a service)
+// then matches that Service to a Deployment by comparing label selectors (if the Service uses label selectors).
+// The names of both Service and Deployment are returned if found.
+func findMatchingDeploymentAndServiceForWebhook(c *collector.Manifests, wcc admissionregv1.WebhookClientConfig) (depName, serviceName string) {
+	// Return if a service reference is not specified, since a URL will be in that case.
+	if wcc.Service == nil {
+		return
+	}
+
+	// Find the matching service, if any. The webhook server may be externally managed
+	// if no service is created by the operator.
+	var ws *corev1.Service
+	for i, service := range c.Services {
+		if service.GetName() == wcc.Service.Name {
+			ws = &c.Services[i]
+			break
+		}
+	}
+	if ws == nil {
+		return
+	}
+	serviceName = ws.GetName()
+
+	// Only ExternalName-type services cannot have selectors.
+	if ws.Spec.Type == corev1.ServiceTypeExternalName {
+		return
+	}
+
+	// If a selector does not exist, there is either an Endpoint or EndpointSlice object accompanying
+	// the service so it should not be added to the CSV.
+	if len(ws.Spec.Selector) == 0 {
+		return
+	}
+
+	// Match service against pod labels, in which the webhook server will be running.
+	for _, dep := range c.Deployments {
+		sel := dep.Spec.Template.GetLabels()
+		// A null label selector matches no objects.
+		if sel == nil {
+			continue
+		}
+		// An empty label selector matches all objects.
+		if len(sel) == 0 {
+			depName = dep.GetName()
+			break
+		}
```

Comment:
Are you looking at the deployment's selector which is `spec.Selector` or the pod labels here?

If it's just the pod labels the var should be renamed to reflect that: `sel := dep.Spec.Template.GetLabels()` to `podLabels := dep.Spec.Template.GetLabels()`
Also the comments saying `label selector`.


Plus what does `An empty label selector matches all objects` mean here?
The label selector can't be empty. It's a required field. But even if it is, what does that signify here.

Reply:
Originally I was using the dep's selector, but that proved unnecessary (and much more difficult) because the webhook service already has a selector. Therefore the variable name and empty selector check are irrelevant, will change/remove.

Reply:
Additionally the label comparison needs to check if `service.Spec.Selector` is a subset of `dep.Spec.Template.GetLabels()`, since the service provides the label constraint here.

Reply:
That makes more sense, thanks for clarifying.